### ISSUE TITLE
[CodeGen] ConvertToDPS pass can't replace tensor.empty ops with flow.tensor.load.* ops

### ISSUE THREAD

**Issue by hanhanW** [iree-org/iree#14316]
The pass intends to replace all tensor.empty ops with destination tensor when it's possible. It can't handle some cases while `--iree-eliminate-empty-tensors` pass can handle it. We might want to leverage the replacement to the new pass.

To repro: `iree-opt --iree-codegen-convert-to-destination-passing-style ~/repro.mlir`

```mlir
#config = #iree_codegen.lowering_config<tile_sizes = [[128, 128, 0], [8, 128, 0], [0, 0, 16], [8, 16, 0], [0, 0, 1]]>
#executable_target_system_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "system-elf-x86_64", {cpu = "cascadelake", cpu_features = "+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-linux-elf", ukernels = false}>
#executable_target_system_elf_x86_64_1 = #hal.executable.target<"llvm-cpu", "system-elf-x86_64", {cpu = "cascadelake", cpu_features = "+avx512f,+avx512dq,+avx512cd,+avx512bw,+avx512vl,+avx,+avx2,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-linux-elf", ukernels = false}>
#map = affine_map<()[s0] -> (s0 * 128)>
#map1 = affine_map<(d0) -> (-d0 + 391, 128)>
#map2 = affine_map<(d0, d1) -> (d1)>
#map3 = affine_map<(d0, d1) -> (d0, d1)>
#pipeline_layout = #hal.pipeline.layout<push_constants = 2, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>
#translation = #iree_codegen.translation_info<CPUDoubleTilingExpert>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_system_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu]} {
  hal.executable public @main_dispatch_22 {
    hal.executable.variant public @system_elf_x86_64, target = #executable_target_system_elf_x86_64_1 {
      hal.executable.export public @main_dispatch_22_matmul_391x384x384_f32 ordinal(0) layout(#pipeline_layout) attributes {translation_info = #translation} {
      ^bb0(%arg0: !hal.device):
        %c3 = arith.constant 3 : index
        %c4 = arith.constant 4 : index
        %c1 = arith.constant 1 : index
        hal.return %c3, %c4, %c1 : index, index, index
      }
      builtin.module {
        func.func @main_dispatch_22_matmul_391x384x384_f32() {
          %c384 = arith.constant 384 : index
          %c391 = arith.constant 391 : index
          %cst = arith.constant 0.000000e+00 : f32
          %cst_0 = arith.constant 6.000000e+00 : f32
          %c600576 = arith.constant 600576 : index
          %c0 = arith.constant 0 : index
          %0 = hal.interface.constant.load[0] : i32
          %1 = hal.interface.constant.load[1] : i32
          %2 = arith.index_castui %0 {stream.alignment = 1024 : index, stream.values = [205824 : index, 795648 : index, 1385472 : index, 1975296 : index, 2565120 : index, 3154944 : index, 3744768 : index]} : i32 to index
          %3 = arith.index_castui %1 {stream.alignment = 1024 : index, stream.values = [0 : index, 3072 : index, 6144 : index, 9216 : index, 12288 : index, 15360 : index, 18432 : index]} : i32 to index
          %4 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<391x384xf32>>
          %5 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%2) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<384x384xf32>>
          %6 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%3) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<384xf32>>
          %7 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c600576) : !flow.dispatch.tensor<writeonly:tensor<391x384xf32>>
          %workgroup_id_x = hal.interface.workgroup.id[0] : index
          %workgroup_count_x = hal.interface.workgroup.count[0] : index
          %workgroup_id_y = hal.interface.workgroup.id[1] : index
          %workgroup_count_y = hal.interface.workgroup.count[1] : index
          %8 = affine.apply #map()[%workgroup_id_y]
          %9 = affine.apply #map()[%workgroup_count_y]
          scf.for %arg0 = %8 to %c391 step %9 {
            %10 = affine.min #map1(%arg0)
            %11 = affine.apply #map()[%workgroup_id_x]
            %12 = affine.apply #map()[%workgroup_count_x]
            scf.for %arg1 = %11 to %c384 step %12 {
              %13 = flow.dispatch.tensor.load %6, offsets = [%arg1], sizes = [128], strides = [1] : !flow.dispatch.tensor<readonly:tensor<384xf32>> -> tensor<128xf32>
              %14 = flow.dispatch.tensor.load %4, offsets = [%arg0, 0], sizes = [%10, 384], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<391x384xf32>> -> tensor<?x384xf32>
              %15 = flow.dispatch.tensor.load %5, offsets = [0, %arg1], sizes = [384, 128], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<384x384xf32>> -> tensor<384x128xf32>
              %16 = tensor.empty(%10) : tensor<?x128xf32>
              %17 = linalg.fill ins(%cst : f32) outs(%16 : tensor<?x128xf32>) -> tensor<?x128xf32>
              %18 = linalg.matmul {lowering_config = #config} ins(%14, %15 : tensor<?x384xf32>, tensor<384x128xf32>) outs(%17 : tensor<?x128xf32>) -> tensor<?x128xf32>
              %19 = linalg.generic {indexing_maps = [#map2, #map3, #map3], iterator_types = ["parallel", "parallel"]} ins(%13, %18 : tensor<128xf32>, tensor<?x128xf32>) outs(%16 : tensor<?x128xf32>) {
              ^bb0(%in: f32, %in_1: f32, %out: f32):
                %20 = arith.addf %in, %in_1 : f32
                %21 = arith.minf %20, %cst_0 : f32
                %22 = arith.maxf %21, %cst : f32
                linalg.yield %22 : f32
              } -> tensor<?x128xf32>
              flow.dispatch.tensor.store %19, %7, offsets = [%arg0, %arg1], sizes = [%10, 128], strides = [1, 1] : tensor<?x128xf32> -> !flow.dispatch.tensor<writeonly:tensor<391x384xf32>>
            }
          }
          return
        }
      }
    }
  }
}
```

**Comment by MaheshRavishankar** [comment]
Thanks for the repro! This should be a simple case for the ConvertToDestinationPassingStyle pass to handle. I agree we should probably use `eliminate-empty-tensors`, but I would put that as a low priority for now cause this can be a a rabbit hole of missed coverage in the upstream pass....

**Comment by MaheshRavishankar** [comment]
I am looking into this, but I think this is purposely not eliminating the `tensor.empty`... I want to see what the final code generated by `iree-eliminate-empty-tensors` pass does

**Comment by hanhanW** [comment]
Below is the output generated by `iree-eliminate-empty-tensors`:

```mlir
#config = #iree_codegen.lowering_config<tile_sizes = [[128, 128, 0], [8, 128, 0], [0, 0, 16], [8, 16, 0], [0, 0, 1]]>
#executable_target_system_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "system-elf-x86_64", {cpu = "cascadelake", cpu_features = "+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-linux-elf", ukernels = false}>
#executable_target_system_elf_x86_64_1 = #hal.executable.target<"llvm-cpu", "system-elf-x86_64", {cpu = "cascadelake", cpu_features = "+avx512f,+avx512dq,+avx512cd,+avx512bw,+avx512vl,+avx,+avx2,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-linux-elf", ukernels = false}>
#map = affine_map<()[s0] -> (s0 * 128)>
#map1 = affine_map<(d0) -> (-d0 + 391, 128)>
#map2 = affine_map<(d0, d1) -> (d1)>
#map3 = affine_map<(d0, d1) -> (d0, d1)>
#pipeline_layout = #hal.pipeline.layout<push_constants = 2, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>
#translation = #iree_codegen.translation_info<CPUDoubleTilingExpert>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_system_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu]} {
  hal.executable public @main_dispatch_22 {
    hal.executable.variant public @system_elf_x86_64, target = #executable_target_system_elf_x86_64_1 {
      hal.executable.export public @main_dispatch_22_matmul_391x384x384_f32 ordinal(0) layout(#pipeline_layout) attributes {translation_info = #translation} {
      ^bb0(%arg0: !hal.device):
        %c3 = arith.constant 3 : index
        %c4 = arith.constant 4 : index
        %c1 = arith.constant 1 : index
        hal.return %c3, %c4, %c1 : index, index, index
      }
      builtin.module {
        func.func @main_dispatch_22_matmul_391x384x384_f32() {
          %c384 = arith.constant 384 : index
          %c391 = arith.constant 391 : index
          %cst = arith.constant 0.000000e+00 : f32
          %cst_0 = arith.constant 6.000000e+00 : f32
          %c600576 = arith.constant 600576 : index
          %c0 = arith.constant 0 : index
          %0 = hal.interface.constant.load[0] : i32
          %1 = hal.interface.constant.load[1] : i32
          %2 = arith.index_castui %0 {stream.alignment = 1024 : index, stream.values = [205824 : index, 795648 : index, 1385472 : index, 1975296 : index, 2565120 : index, 3154944 : index, 3744768 : index]} : i32 to index
          %3 = arith.index_castui %1 {stream.alignment = 1024 : index, stream.values = [0 : index, 3072 : index, 6144 : index, 9216 : index, 12288 : index, 15360 : index, 18432 : index]} : i32 to index
          %4 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<391x384xf32>>
          %5 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%2) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<384x384xf32>>
          %6 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%3) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<384xf32>>
          %7 = hal.interface.binding.subspan set(0) binding(2) type(storage_buffer) alignment(64) offset(%c600576) : !flow.dispatch.tensor<writeonly:tensor<391x384xf32>>
          %workgroup_id_x = hal.interface.workgroup.id[0] : index
          %workgroup_count_x = hal.interface.workgroup.count[0] : index
          %workgroup_id_y = hal.interface.workgroup.id[1] : index
          %workgroup_count_y = hal.interface.workgroup.count[1] : index
          %8 = affine.apply #map()[%workgroup_id_y]
          %9 = affine.apply #map()[%workgroup_count_y]
          scf.for %arg0 = %8 to %c391 step %9 {
            %10 = affine.min #map1(%arg0)
            %11 = affine.apply #map()[%workgroup_id_x]
            %12 = affine.apply #map()[%workgroup_count_x]
            scf.for %arg1 = %11 to %c384 step %12 {
              %13 = flow.dispatch.tensor.load %6, offsets = [%arg1], sizes = [128], strides = [1] : !flow.dispatch.tensor<readonly:tensor<384xf32>> -> tensor<128xf32>
              %14 = flow.dispatch.tensor.load %4, offsets = [%arg0, 0], sizes = [%10, 384], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<391x384xf32>> -> tensor<?x384xf32>
              %15 = flow.dispatch.tensor.load %5, offsets = [0, %arg1], sizes = [384, 128], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<384x384xf32>> -> tensor<384x128xf32>
              %16 = flow.dispatch.tensor.load %7, offsets = [%arg0, %arg1], sizes = [%10, 128], strides = [1, 1] : !flow.dispatch.tensor<writeonly:tensor<391x384xf32>> -> tensor<?x128xf32>
              %17 = linalg.fill ins(%cst : f32) outs(%16 : tensor<?x128xf32>) -> tensor<?x128xf32>
              %18 = linalg.matmul {lowering_config = #config} ins(%14, %15 : tensor<?x384xf32>, tensor<384x128xf32>) outs(%17 : tensor<?x128xf32>) -> tensor<?x128xf32>
              %19 = linalg.generic {indexing_maps = [#map2, #map3, #map3], iterator_types = ["parallel", "parallel"]} ins(%13, %18 : tensor<128xf32>, tensor<?x128xf32>) outs(%16 : tensor<?x128xf32>) {
              ^bb0(%in: f32, %in_1: f32, %out: f32):
                %20 = arith.addf %in, %in_1 : f32
                %21 = arith.minf %20, %cst_0 : f32
                %22 = arith.maxf %21, %cst : f32
                linalg.yield %22 : f32
              } -> tensor<?x128xf32>
              flow.dispatch.tensor.store %19, %7, offsets = [%arg0, %arg1], sizes = [%10, 128], strides = [1, 1] : tensor<?x128xf32> -> !flow.dispatch.tensor<writeonly:tensor<391x384xf32>>
            }
          }
          return
        }
      }
    }
  }
}

```

**Comment by MaheshRavishankar** [comment]
Ok cool. It is good that the upstream pass is doing this, it was always an ask. Would it be useful for me to look at this further and fix `ConvertToDestinationPassingStyle` to do this?

**Comment by hanhanW** [comment]
That would be very useful and will unblock @dcaballe multi level tiling work. We can't simply use `iree-eliminate-empty-tensors` here because it will convert a tensor.pad op into other operations. We want to keep the pad op during tiling and vectorization. So it's better if we can teach `ConvertToDestinationPassingStyle` to do this.

**Comment by benvanik** [comment]
(closing as stale)
